### PR TITLE
refactor(registration): refactor registration viewset with exceptions

### LINKED ISSUE #152: Innfør exceptions for å kontrollere API-ets error responses
Koden vår kan dra stor nytte av egendefinerte unntak som kan returneres som en API respons ved feil. På den måten får vi standardisert API responsen ved feil og legge bedre til rette for gjenbruk av kode.
Det reduserer mengden kode og øker dermed lesbarheten ved vi kan skrive
`raise EventNotFound`
istedenfor 
`return Response(detail=_("Event could not be found", status=404)`

Noen av unntakene kan for eksempel arve fra Django REST sin klasse `APIException` og for eksempel inneholde feltene `detail` og `status_code`

Dette innebærer også å konvertere python og Django sine unntak til restframework sine unntak, da sistnevnte sørger for at vi gir en skikkelig respons med status kode til frontend (og ikke bare 500). En guide finnes her: https://github.com/HackSoftware/Django-Styleguide#handle-exceptions-in-apis 

Her er det viktig at vi utarbeider en standard for implementasjon og bruk av unntak og navnekonvensjoner. Det bør diskuteres i backend.

Mer info:
https://stackoverflow.com/a/10890416 
https://www.django-rest-framework.org/api-guide/exceptions/#custom-exception-handling
https://stackoverflow.com/a/62586259

### PR DESCRIPTION
**Pull request by Eirsteir**
## Add custom exception handling, an error mixin which converts regular exceptions to DRF exceptions and a handler which logs all api errors. Simplified the viewset by splitting methods and using more of DRF built in logic.
Link to Trello task: closes #152 

Comments/issues:

The exceptions will be converted to one of Django Rest Framework's exceptions which automatically return a Response object with the default detail and status code when raised. This allows us to define custom error messages and status codes on the exceptions without having to catch them manually. 

It also allows for much cleaner and simpler viewsets by removing the need for (almost) all exception handling, since the default return detail message is automatically included in the response.

The exceptions that are not catched are also logged after being handled by DRF's default handler in the custom exception handler.

The base mixin class is like a decorator for the viewset classes which simply add extra behaviour to the viewset, in this case convert known exceptions to DRF exceptions defined in the spesific subclass, such as `APIRegistrationErrorMixin`. I have used this here to specify some Registration-spesific exceptions to avoid cluttering the base error mixin with exception types from all of the models.

It retrieves the message from the exception and sends it to the new one. 

The `expected_exceptions` properties is a translator between Django/pythons exceptions which we can add new exceptions to which do not extend one of DRF's exceptions.

## Pull request checklist

Please check if your PR fulfills the following requirements:
- [x] PR name is descriptive
- [x] The PR includes the link to the relevant Trello task
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been reviewed and added / updated if needed (for bug fixes / features)

- [x] `python manage.py runserver` was run locally witout errors

## Pull request type

Please check the type of change your PR introduces:
- [ ] Bugfix
- [ ] Feature
- [x] Code style update (formatting, renaming)
- [x] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation content changes
- [ ] Other (please describe):

### CODE REVIEW

File: app/content/views/registration.py [hunk 1]
```diff
@@ -1,141 +1,103 @@
...
+                registration.event.title,
+                [registration.user.email],
             )
+
+    def destroy(self, request, *args, **kwargs):
+        registration = self.get_object()
+        if is_admin_user(request):
+            return self._admin_unregister(registration)
+
+        if self._attempts_to_unregister_self():
```
Comment (norgor) [inline app/content/views/registration.py:88]: This should become a guard clause such as:
```
if self._is_not_own_registration():
    raise PermissionDenied("You can only unregister yourself")

return self._unregister(registration)
```
The reason is that this should be a special case that is clearly handled on it's own.

File: app/content/views/registration.py [hunk 2]
```diff
@@ -1,141 +1,103 @@
...
-            return Response(
-                {"detail": _("The provided event does not exist")}, status=404
+    def update(self, request, *args, **kwargs):
+        registration = self.get_object()
+        serializer = self.get_serializer(data=request.data)
+        serializer.is_valid(raise_exception=True)
+
+        # TODO: what is the purpose of this?
+        if self._user_has_already_attended_event(registration):
+            raise APIUserAlreadyAttendedEvent
```
Comment (norgor) [inline app/content/views/registration.py:66]: did you forget to construct the class here?

**Review by norgor — CHANGES_REQUESTED** [review]
I really love how it looks so far! This is going to be super smooth! 😎I noted some things that could or should be changed. Nice work! 🥇

File: app/content/models/event.py [hunk 3]
```diff
@@ -71,6 +71,10 @@ def waiting_list_count(self):
         """ Number of users on the waiting list """
         return Registration.objects.filter(event__pk=self.pk, is_on_wait=True).count()
 
+    @property
+    def sign_off_deadline_has_passed(self):
+        return self.sign_off_deadline < today()
```
[the remark this answers is not in the document]
Comment (Eirsteir) [reply app/content/models/event.py:76]: I totally agree with you on this, thanks for pointing it out 😄

File: app/content/views/registration.py [hunk 4]
```diff
@@ -1,141 +1,103 @@
...
-        except ValidationError as e:
-            return Response({"detail": _(e.message)}, status=404)
-
-    def destroy(self, request, event_id, user_id):
-        """ Unregisters the user specified with provided event_id and user_id """
-        try:
-            registration = Registration.objects.get(
-                event__pk=event_id, user__user_id=user_id
-            )
+    def _attempts_to_unregister_self(self):
```
[the remark this answers is not in the document]
Comment (Eirsteir) [reply app/content/views/registration.py:37]: I was a bit unsure of this name myself, but I like `is_own_registration`. It's clearer 😊

File: app/content/views/registration.py (hunk 2 above)
Comment (Eirsteir) [reply app/content/views/registration.py:66]: It is a matter of semantics in Python as the exception object is implicitly created (or `()` is added with other words), but the fact that you asked shows that should be with the parantheses 😉

File: app/content/views/registration.py [hunk 5]
```diff
@@ -1,141 +1,103 @@
...
-        except Event.DoesNotExist:
-            return Response(
-                {"detail": _("The provided event does not exist")}, status=404
+    def update(self, request, *args, **kwargs):
+        registration = self.get_object()
+        serializer = self.get_serializer(data=request.data)
+        serializer.is_valid(raise_exception=True)
+
+        # TODO: what is the purpose of this?
+        if self._user_has_already_attended_event(registration):
```
Comment (Zenjjim) [inline app/content/views/registration.py:65]: Brukerstory: Hvis en har ankommet et arrangement, og en annen prøver å komme inn med samme QR kode, skal det bli kastet en feil som sier "brukerene har allerede ankommet"

**Review by Zenjjim — APPROVED** [review]

File: app/content/views/registration.py (hunk 5 above)
Comment (Eirsteir) [reply app/content/views/registration.py:64]: Takk, legger inn en liten kommentar siden den har en spesiell betydning

File: app/content/views/registration.py (hunk 2 above)
Comment (norgor) [reply app/content/views/registration.py]: Aaah! Didn't know that! I guess raise will construct it for you then. There was a PEP proposal for it in 2003, but argument against it was that current code had to be changed. [https://www.python.org/dev/peps/pep-0317/](url)

**Review by norgor — COMMENTED** [review]
Great job! 🥇

**Review by norgor — APPROVED** [review]
Great job! 🥇

**Pull request merged**